### PR TITLE
Fix Azure Batch netcore header signing

### PR DESCRIPTION
**Pull request by matthchr**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

**Comment by matthchr** [comment]
This fixes issue #3170

**Comment by matthchr** [comment]
Please **DO NOT MERGE** until I get signoff from my team members: @itowlson @darylmsft @xingwu1 @kiwidev 

I will respond once we are ready to merge.

**Comment by shahabhijeet** [comment]
@matthchr please remove "Do not merge" label and add need-review, so that I know it is ready to be reviewed.

**Comment by matthchr** [comment]
@shahabhijeet Thanks, didn't realize we had those labels. In the future I can add them myself as well.

I'll let you know when we're ready to merge.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (darylmsft) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: are these methods usually named async Task XxxYyyAsync()?

**Review by darylmsft — APPROVED** [review]

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (kiwidev) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Normal convention is not to add Async suffix to tests - they are leaf methods and never called by other places. I'm unsure if there is other prior art in this project though.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (kiwidev) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Do we need to iterate through ports until we find a free one?

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (kiwidev) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: While these tests are good, they test an aspect of the underlying behavior, but don't seem to actually test our shared key algorithm. What we're wanting to do is enforce that the StringToSign that we computed (made up of headers etc) matches the StringToSign that would be computed by the server (based on the headers sent).
So we'd either need a way of extracting the StringToSign from our BatchSharedKeyCredential (do we expose this?) or else compute the hash again on the server side and validate that they match.

I think we should keep these tests too as a way of exploring the underlying functionality / confirming that different platforms don't change behavior, however the "SharedKeyAuth" doesn't seem to belong inside this test as it isn't being asserted against.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (kiwidev) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Patch?
Options?

File: src/SDKs/Batch/DataPlane/Azure.Batch/Protocol/BatchSharedKeyCredential.cs
Comment (kiwidev) [inline src/SDKs/Batch/DataPlane/Azure.Batch/Protocol/BatchSharedKeyCredential.cs]: I'd still prefer that the `#if FullNetFx` was outside of the if statement so we're not generating an empty if for dotNetCore.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (kiwidev) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Assert that it's equal to 0?

**Review by kiwidev — COMMENTED** [review]
I think the logic is ok, comments are related to the tests.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (itowlson) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: +1 - the Async suffix is a hint to human callers of a method, which is not an issue for tests

Comment (itowlson) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Test name does not describe what is being asserted/tested.  Reading the code this one seems to contain a number of assertions, not all of which relate to content length; and the content length assertions vary from HTTP method to HTTP method.  So I'm not sure what to suggest as a better name except that it should be a statement which can be true or false.  (You might want to consider breaking out the different HTTP methods so you could have more specific test names for each, e.g. `HttpDelete_WithSharedKeySigning_DoesNotIncludeContentLength`.)

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (itowlson) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: This function name doesn't mean a whole lot to me.  'Lambda' seems to describe more its intended usage than its actual semantics...?

Comment (itowlson) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Putting this long local function in the middle of test impacts readability.  Consider restructuring to keep the test readable top-to-bottom.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (itowlson) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Is HttpMethod an enum?  If so consider using a switch instead of a series of if-clauses.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (matthchr) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: The tests in these projects don't mention `Async` (even if they are async). Leaving this as is.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (matthchr) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Agree with basically all you said.

> So we'd either need a way of extracting the StringToSign from our BatchSharedKeyCredential (do we expose this?) or else compute the hash again on the server side and validate that they match.

Yes, I have another item to enable the end to end integration tests to run in both netcore and netframework, but in order to do that there is more work required. So instead of doing that for this issue (since it's customer impacting), I added a targeted unit test. We will circle back around and enable the end to end tests to run against both, at which point we should get the validation we seek (since the E2E tests use all the different verbs)

Comment (matthchr) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: I'll fix the name of the test class/test to more clearly describe what it is testing

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (matthchr) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Daryl brought this up too... even though we don't use `Options` anywhere internally, I guess I'll add it.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (matthchr) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Sadly it's not an enum

File: src/SDKs/Batch/DataPlane/Azure.Batch/Protocol/BatchSharedKeyCredential.cs
Comment (matthchr) [inline src/SDKs/Batch/DataPlane/Azure.Batch/Protocol/BatchSharedKeyCredential.cs]: Alright - I've made that change

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (matthchr) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Fixed I think.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (matthchr) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: The issue with a line in the sand name like "DoesNot" or "Does" with respect to including content length is that the behavior varies across target framework, so unless I conditionally compile the whole test out, it's going to be wrong for one of the frameworks when it comes to `DELETE`

I could add a bunch of tests clearly spelling out each case (either does or does not), but they would all just call into a common helper function (which is the whole point of `Theory`).

I've changed this to be: `HttpClient_IncludesContentLengthHeaderOnExpectedHttpVerbs`. Obviously what is "Expected" is still nebulous (you have to read the test code)...

Let me know if you don't think this is sufficient.

File: src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs
Comment (matthchr) [inline src/SDKs/Batch/DataPlane/Azure.Batch.Unit.Tests/BatchSharedKeyCredentialUnitTests.cs]: Fixed

**Review by itowlson — APPROVED** [review]

**Comment by matthchr** [comment]
@shahabhijeet - Okay, this needs review from an SDK team member and merge now.

**Pull request merged**
